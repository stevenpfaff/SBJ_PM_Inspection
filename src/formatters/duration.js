const MINS_IN_HOUR = 60

const formatHours = hours => {
  if (typeof hours !== 'number' || Number.isNaN(hours))
    throw new Error(
      `Invalid hours value: received '${hours}' - expected number`
    )

  const wholeHours = Math.floor(hours)

  if (wholeHours === 1) {
    return `${wholeHours} hr`
  }

  return `${wholeHours} hrs`
}

const formatMinutes = hours => {
  if (typeof hours !== 'number' || Number.isNaN(hours))
    throw new Error(
      `Invalid hours value: received '${hours}' - expected number`
    )

  const minutes = hours * MINS_IN_HOUR

  const wholeMinutes = Math.floor(minutes % MINS_IN_HOUR)

  if (wholeMinutes) {
    if (wholeMinutes === 1) {
      return `${wholeMinutes} min`
    }

    return `${wholeMinutes} mins`
  }

  return ''
}

const duration = {
  formatHours,
  formatMinutes
}

export default duration
