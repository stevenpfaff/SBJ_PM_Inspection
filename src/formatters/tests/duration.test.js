import duration from '../duration'

describe('duration', () => {
  describe('formatHours', () => {
    it('should throw an error if typeof hours is not a number', () => {
      let actual
      let error

      try {
        actual = duration.formatHours()
      } catch (e) {
        error = e
      }

      expect(actual).toBeUndefined()
      expect(error).toMatchSnapshot()
    })

    it('should throw an error if hours is not a number', () => {
      let actual
      let error

      try {
        actual = duration.formatHours(NaN)
      } catch (e) {
        error = e
      }

      expect(actual).toBeUndefined()
      expect(error).toMatchSnapshot()
    })

    it('should return "1 hr" if hours === 1', () => {
      const expected = '1 hr'

      const actual = duration.formatHours(1)

      expect(actual).toBe(expected)
    })

    it('should return "1 hr" if 1 < hours && hours > 2', () => {
      const expected = '1 hr'

      const actual = duration.formatHours(1.3)

      expect(actual).toBe(expected)
    })

    it('should return formatted hours if hours > 1', () => {
      const expected = '2 hrs'

      const actual = duration.formatHours(2)

      expect(actual).toBe(expected)
    })

    it('should return formatted hours if hours is greater than 2 and not a whole number', () => {
      const expected = '2 hrs'

      const actual = duration.formatHours(2.5)

      expect(actual).toBe(expected)
    })
  })

  describe('formatMinutes', () => {
    it('should throw an error if typeof hours is not a number', () => {
      let actual
      let error

      try {
        actual = duration.formatMinutes()
      } catch (e) {
        error = e
      }

      expect(actual).toBeUndefined()
      expect(error).toMatchSnapshot()
    })

    it('should throw an error if hours is not a number', () => {
      let actual
      let error

      try {
        actual = duration.formatMinutes(NaN)
      } catch (e) {
        error = e
      }

      expect(actual).toBeUndefined()
      expect(error).toMatchSnapshot()
    })

    it('should return an empty string if hours is a whole number', () => {
      const expected = ''

      const actual = duration.formatMinutes(1)

      expect(actual).toBe(expected)
    })

    it('should return "1 min" if hour ends with ".0167"', () => {
      const expected = '1 min'

      const actual = duration.formatMinutes(0.0167)

      expect(actual).toBe(expected)
    })

    it('should return formatted minutes if 1 < hours && hours > 2', () => {
      const expected = '18 mins'

      const actual = duration.formatMinutes(1.3)

      expect(actual).toBe(expected)
    })

    it('should return formatted hours if hours is greater than 2 and not a whole number', () => {
      const expected = '30 mins'

      const actual = duration.formatMinutes(2.5)

      expect(actual).toBe(expected)
    })
  })
})
