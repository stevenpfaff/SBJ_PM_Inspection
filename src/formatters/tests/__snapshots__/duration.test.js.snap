// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`duration formatHours should throw an error if hours is not a number 1`] = `[Error: Invalid hours value: received 'NaN' - expected number]`;

exports[`duration formatHours should throw an error if typeof hours is not a number 1`] = `[Error: Invalid hours value: received 'undefined' - expected number]`;

exports[`duration formatMinutes should throw an error if hours is not a number 1`] = `[Error: Invalid hours value: received 'NaN' - expected number]`;

exports[`duration formatMinutes should throw an error if typeof hours is not a number 1`] = `[Error: Invalid hours value: received 'undefined' - expected number]`;
